### DISCUSSION TITLE
Conditional resolving of fields

### DISCUSSION THREAD

**Discussion by pylipp** [mirumee/ariadne#678] [Q&A] [answered]
Say I have
```
type Author {
  id: ID
  name: String
  books: [Book]
}
type Book {
  id: ID
  title: String
  author: Author
}
type Query {
  author(id: ID): Author
}
```
and a resolver implementation
```
query = QueryType()

@query.field("author")
def resolve(obj, info, id):
    data = orm.get_author(id).as_dict()
    data["books"] = orm.get_books_from_author(id).as_dict()
    return data
```
I would overfetch when doing a query like
```
query {
  author(id: "1") {
    name
  }
}
```
How can I work around this? Do I have access to the requested fields in the `resolve` function?

**Comment by rafalp** [comment, marked as answer]
> How can I work around this?

Do you have a measurable data showing that this is degrading performance of your app?

You can look into requested fields from author resolver by looking up the contents of `info` instance, but overhead of python logic introspecting this object and mapping it to your ORM would be more than gains from limiting number of columns in your SQL query.

**Reply by pylipp** [reply]
Thank you for the quick reply!
> > How can I work around this?
> 
> Do you have a measurable data showing that this is degrading performance of your app?

I don't have data supporting that claim. It's something I assumed people are trying to avoid (don't query unnecessarily).
> 
> You can look into requested fields from author resolver by looking up the contents of `info` instance, but overhead of python logic introspecting this object and mapping it to your ORM would be more than gains from limiting number of columns in your SQL query.

Okay I see. It's good to know that the `info` introspection is costly. (Technically I think it's not the number of columns I'm trying to save but a complete second DB query call. First `orm` function above queries the authors table, second queries the books table)

**Reply by pylipp** [reply]
I think I was a bit confused here. I can write another resolver
```
author = ObjectType("Author")

@author.field("books")
def resolve_author_books(author_obj, info):
    return orm.get_books_from_author(author_obj.id)
```
The query resolver simplifies to

```
@query.field("author")
def resolve(obj, info, id):
    return orm.get_author(id)
```

and I can be sure that the `books` field will only be resolved when specified in the query body.
